### PR TITLE
mgr/dashboard: "Create Cluster" workflow (e2e tests)

### PR DESCRIPTION
**Pull request by avanthakkar**
Fixes: https://tracker.ceph.com/issues/50336
Signed-off-by: Avan Thakkar <athakkar@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by avanthakkar** [comment]
jenkins test make check

**Review by epuertat — COMMENTED** [review]
Now we're adopting TDD (actually BDD or ATDD, since this is end-to-end user-experience testing), I wonder if adopting a better way to express our tests as 'Acceptance criteria', would also help us align both the feature requirements, the test flow and the feature implementation.

Would you like exploring the idea of rewriting this test in 'Gherkin' language. There's [a plugin for Cypress](https://github.com/TheBrainFamily/cypress-cucumber-preprocessor) that allows you doing that: [[tutorial]](https://levelup.gitconnected.com/using-cypress-with-cucumber-to-test-your-app-as-if-you-were-a-user-15288900a1a6) [[another tutorial]](https://wanago.io/2020/01/13/javascript-testing-cypress-cucumber/)

**Comment by alfonsomthd** [comment]
@epuertat @nizamial09 @avanthakkar Maybe I'm missing something here but if you go with TDD, the tests should be in the feature branch (https://github.com/ceph/ceph/pull/41720) so they fail and then the code that make them pass is provided in the same branch.

**Comment by epuertat** [comment]
> @epuertat @nizamial09 @avanthakkar Maybe I'm missing something here but if you go with TDD, the tests should be in the feature branch (#41720) so they fail and then the code that makes them pass is provided in the same branch.

Now we have (or soon to) full CI on feature branches, so I suggest taking that approach.

File: src/pybind/mgr/dashboard/frontend/cypress/integration/cluster/cluster-welcome-page.po.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  }
+
+  doSkip() {
+    cy.get('[name=continue-cluster-creation]').click();
+
+    cy.get('cd-create-cluster').should('exist');
+
+    cy.get('[name=skip]').click();
+    cy.get('cd-dashboard').should('exist');
+  }
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/cypress/integration/cluster/cluster-welcome-page.po.ts:22]:
```suggestion
doLogin() {
    cy.get('[name=username]').type('admin');
    cy.get('#password').type('admin');
    cy.get('[type=submit]').click();
    cy.get('cd-create-cluster').should('exist');
  }

  createCluster() {
    cy.get('[name=create-cluster]').click();

    cy.get('cd-dashboard').should('exist');
    const notification = new NotificationSidebarPageHelper();
    notification.open();
    notification.getNotifications().should('include.text', 'Cluster creation feature not implemented');
  }

  doSkip() {
    cy.get('[name=skip-create-cluster]').click();
    cy.get('cd-dashboard').should('exist');
  }

File: src/pybind/mgr/dashboard/frontend/cypress/integration/orchestrator/06-cluster-welcome-page.e2e-spec.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { CreateClusterWelcomePageHelper } from '../cluster/cluster-welcome-page.po';
+
+describe('Create cluster page', () => {
+  const createCluster = new CreateClusterWelcomePageHelper();
+
+  beforeEach(() => {
+    cy.login();
+    Cypress.Cookies.preserveOnce('token');
+  });
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/cypress/integration/orchestrator/06-cluster-welcome-page.e2e-spec.ts:9]:
```suggestion
  const createCluster = new CreateClusterWelcomePageHelper();
  const loginPage = new LoginPageHelper();

  beforeEach(() => {
    loginPage.navigateTo();
    createCluster.doLogin();
  });

**Review by nizamial09 — CHANGES_REQUESTED** [review]
Since we are going to test what the user is going to see, let's do it like this.
**For Create Cluster**
User Login --> Sees Welcome Page --> Click Create Cluster --> It Fails --> Checks the notification

**For Skip**
User Login --> Sees Welcome Page --> Skip --> Expects landing page.

**Comment by avanthakkar** [comment]
Test passing locally with kcli cluster:
```
 Running:  orchestrator/06-cluster-welcome-page.e2e-spec.ts                                (1 of 1)


  Create cluster page
    ✓ should fail to create cluster (3334ms)
    ✓ should skip to dashboard landing page (1460ms)


  2 passing (5s)


  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        2                                                                                │
  │ Passing:      2                                                                                │
  │ Failing:      0                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  0                                                                                │
  │ Video:        false                                                                            │
  │ Duration:     4 seconds                                                                        │
  │ Spec Ran:     orchestrator/06-cluster-welcome-page.e2e-spec.ts                                 │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


====================================================================================================

  (Run Finished)


       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✔  orchestrator/06-cluster-welcome-pag      00:04        2        2        -        -        - │
  │    e.e2e-spec.ts                                                                               │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✔  All specs passed!                        00:04        2        2        -        -        -  
    ```

**Review by nizamial09 — APPROVED** [review]

**Review by alfonsomthd — APPROVED** [review]

**Review by epuertat — APPROVED** [review]

**Pull request merged**